cpp/src/strings/convert/convert_urls.cu

```diff
@@ -0,0 +1,279 @@
+                                    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                    cudaStream_t stream = 0)
+{
+    size_type strings_count = strings.size();
+    if( strings_count == 0 )
+        return make_empty_strings_column(mr,stream);
+
+    auto strings_column = column_device_view::create(strings.parent(), stream);
+    auto d_strings = *strings_column;
+
+    // copy null mask
+    rmm::device_buffer null_mask = copy_bitmask(strings.parent(),stream,mr);
+    // build offsets column
+    auto offsets_transformer_itr = thrust::make_transform_iterator( thrust::make_counting_iterator<size_type>(0),
+                                                                    url_encoder_fn{d_strings} );
+    auto offsets_column = make_offsets_child_column(offsets_transformer_itr,
+                                                    offsets_transformer_itr+strings_count,
+                                                    mr, stream);
+    auto d_offsets = offsets_column->view().data<int32_t>();
+    // build chars column
+    auto chars_column = create_chars_child_column( strings_count, strings.null_count(),
+                                                   thrust::device_pointer_cast(d_offsets)[strings_count],
+                                                   mr, stream );
+    auto d_chars = chars_column->mutable_view().data<char>();
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+                       thrust::make_counting_iterator<size_type>(0), strings_count,
+                       url_encoder_fn{d_strings,d_offsets,d_chars});
+    //
+    return make_strings_column(strings_count, std::move(offsets_column), std::move(chars_column),
+                               strings.null_count(), std::move(null_mask), stream, mr);
+}
+
+} // namespace detail
+
+// external API
+std::unique_ptr<column> url_encode( strings_column_view const& strings,
+                                    rmm::mr::device_memory_resource* mr)
+{
+    return detail::url_encode(strings,mr);
+}
+
+namespace detail
+{
+namespace
+{
+
+//
+// This is the functor for the url_decode() method below.
+// Specific requirements are documented in custrings issue #321.
+// In summary it converts all character sequences starting with '%' into bytes
+// interpretting the following 2 characters as hex values to create the output byte.
+// For example, the sequence '%20' is converted into byte (0x20) which is a single
+// space character. Another example converts '%C3%A9' into 2 sequential bytes
+// (0xc3 and 0xa9 respectively). Overall, 3 characters are converted into one byte
+// whenever a '%' character is encountered in the string.
+//
+struct url_decoder_fn
+{
+    column_device_view const d_strings;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    // utility to convert 2 hex chars into a single byte
+    __device__ char hex_to_byte( char ch1, char ch2 )
+    {
+        u_char result = 0;
```

Comment:
The code below is repeated twice for `ch1` and `ch2`. It might be a good idea to factor it out and then the resulting code should just read something like:
```cpp
u_char result = 16 * fn(ch1) + fn(ch2);
return static_cast<char>(result);
```
Where `fn` is the name of the factored out function